src/Microsoft.Azure.WebJobs.Host/Queues/Listeners/QueueListener.cs

```diff
@@ -244,13 +244,23 @@ private TaskSeriesCommandResult CreateSucceededResult()
 
         internal async Task ProcessMessageAsync(IStorageQueueMessage message, TimeSpan visibilityTimeout, CancellationToken cancellationToken)
         {
+            // These values may change if the message is inserted into another queue. We'll store them here and make sure
+            // the message always has the original values before we pass it to a customer-facing method.
+            string id = message.Id;
+            string popReceipt = message.PopReceipt;
+            DateTimeOffset? insertionTime = message.InsertionTime;
+            DateTimeOffset? nextVisibleTime = message.NextVisibleTime;
+            DateTimeOffset? expirationTime = message.ExpirationTime;
+
             try
             {
                 if (!await _queueProcessor.BeginProcessingMessageAsync(message.SdkObject, cancellationToken))
                 {
                     return;
                 }
 
+                message.SdkObject.SetProperties(id, popReceipt, insertionTime, nextVisibleTime, expirationTime);
```

Comment:
I wonder if we shouldn't just scope this to CopyMessageToPoisonQueue below to minimize impact. I know you've done this for completeness, but I think we'd be fine just scoping the fix to that?